schema/encoding/jsonschema/schema.go

```diff
@@ -18,132 +13,104 @@ var (
 	ErrNotImplemented = errors.New("not implemented")
 )
 
-// encodeSchema writes JSON Schema keys and values based on s, without the outer curly braces, to w.
-func encodeSchema(w io.Writer, s *schema.Schema) (err error) {
+const (
+	// schemaSizeHint is the maximum number of keys that can be added to a schema map based on schema.Schema.
+	schemaSizeHint = 7
+	// fieldSizeHint is the maximum number of keys that can be added to a schema map based on schema.Field
+	// attributes excluding the Validator.
+	fieldSizeHint = 3
+)
+
+func formatSchema(m map[string]interface{}, s *schema.Schema) (err error) {
 	if s == nil {
 		return
 	}
-
-	ew := errWriter{w: w}
+	m["type"] = "object"
 	if s.Description != "" {
-		ew.writeFormat(`"description": %q, `, s.Description)
-	}
-	ew.writeString(`"type": "object", `)
-	ew.writeString(`"additionalProperties": false, `)
-	ew.writeString(`"properties": {`)
-	var required []string
-	var notFirst bool
-	for _, key := range sortedFieldNames(s.Fields) {
-		field := s.Fields[key]
-		if notFirst {
-			ew.writeString(", ")
-		}
-		notFirst = true
-		if field.Required {
-			required = append(required, fmt.Sprintf("%q", key))
-		}
-		ew.err = encodeField(ew, key, field)
-		if ew.err != nil {
-			return ew.err
-		}
+		m["description"] = s.Description
 	}
-	ew.writeString("}")
+	m["additionalProperties"] = false
 	if s.MinLen > 0 {
-		ew.writeFormat(`, "minProperties": %s`, strconv.FormatInt(int64(s.MinLen), 10))
+		m["minProperties"] = s.MinLen
 	}
 	if s.MaxLen > 0 {
-		ew.writeFormat(`, "maxProperties": %s`, strconv.FormatInt(int64(s.MaxLen), 10))
+		m["maxProperties"] = s.MaxLen
 	}
-
-	if len(required) > 0 {
-		ew.writeFormat(`, "required": [%s]`, strings.Join(required, ", "))
+	if len(s.Fields) > 0 {
+		err = formatFields(m, s.Fields)
 	}
-	return ew.err
-}
 
-type fieldWriter struct {
-	errWriter
-	propertiesCount int
+	return err
 }
 
-// comma optionally outputs a comma.  Invoke this when you're about to write a property.  Tracks how many have been
-// written and emits if not the first.
-func (fw *fieldWriter) comma() {
-	if fw.propertiesCount > 0 {
-		fw.writeString(",")
-	}
-	fw.propertiesCount++
-}
+func formatFields(m map[string]interface{}, fields schema.Fields) error {
+	props := make(map[string]interface{}, len(fields))
+	required := make([]string, 0, len(fields))
 
-func (fw *fieldWriter) resetPropertiesCount() {
-	fw.propertiesCount = 0
-}
+	for key, field := range fields {
+		if field.Required {
+			required = append(required, key)
+		}
+		formatter, err := validatorFormatter(field.Validator)
+		if err != nil {
+			return err
+		}
+		fm, err := formatter.JSONSchema()
+		if err != nil {
+			return err
+		}
+		formatField(fm, field)
```

Comment:
If we come up with a new name, we should probably change that consistently, e.g. also for formatSchema.

I am not sure encodeFields is better though, as it's no longer encoding data. I will try to think of something...